File: src/QuadrupedRobot.cpp
#include "QuadrupedRobot.h"
#include <math.h>

QuadrupedRobot::QuadrupedRobot() : 
    pwm(), mpu(), moving(false), currentGait(TROT), maxSpeed(1.0f), bodyHeight(100.0f),
    state(LEARNING) {
    
    nn = new NeuralNetwork(14, 20, 8);  // 14 inputs, 20 hidden, 8 outputs
    
    gaitParams = {
        .stepHeight = 30.0f,
        .strideLength = 50.0f,
        .cycleTime = 1000.0f,
        .phaseOffsets = {0.0f, 0.25f, 0.5f, 0.75f}
    };
    
    initServoConfigs();
    initLegConfigs();
}

QuadrupedRobot::~QuadrupedRobot() {
    delete nn;
}

void QuadrupedRobot::init() {
    Wire.begin();
    pwm.begin();
    pwm.setPWMFreq(50);  // Standard servo frequency
    
    if (!mpu.begin()) {
        Serial.println("MPU6050 initialization failed!");
        return;
    }
    
    mpu.setAccelerometerRange(MPU6050_RANGE_8_G);
    mpu.setGyroRange(MPU6050_RANGE_500_DEG);
    mpu.setFilterBandwidth(MPU6050_BAND_21_HZ);
    
    calibrateSensors();
    stand();
}

void QuadrupedRobot::initServoConfigs() {
    // Configure each servo with mirroring support
    for (int i = 0; i < TOTAL_SERVOS; i++) {
        servoConfigs[i] = {
            .pin = i,
            .minPulse = 150,
            .maxPulse = 600,
            .minAngle = -90.0f,
            .maxAngle = 90.0f,
            .offset = 0.0f,
            .reversed = false,
            .isMirrored = (i % 2 == 1)  // Mirror every second servo
        };
    }
}

void QuadrupedRobot::initLegConfigs() {
    // Initialize leg configurations with proper mirroring
    const float X_OFFSET = 100.0f;
    const float Y_OFFSET = 100.0f;
    
    legConfigs[FRONT_RIGHT] = {X_OFFSET, Y_OFFSET, -bodyHeight, false, 0.0f};
    legConfigs[FRONT_LEFT] = {X_OFFSET, -Y_OFFSET, -bodyHeight, true, 0.0f};
    legConfigs[BACK_RIGHT] = {-X_OFFSET, Y_OFFSET, -bodyHeight, false, 0.0f};
    legConfigs[BACK_LEFT] = {-X_OFFSET, -Y_OFFSET, -bodyHeight, true, 0.0f};
    
    for (int i = 0; i < NUM_LEGS; i++) {
        isGrounded[i] = true;
        lastContactTime[i] = millis();
    }
}

void QuadrupedRobot::update() {
    updateMpuData();
    
    // Basic balance and movement updates
    if (moving) {
        balance();
    }
    
    validateMovement();
    optimizeEnergyUsage();
}

void QuadrupedRobot::calibrateSensors() {
    Serial.println("Calibrating sensors...");
    float accelBias[3] = {0, 0, 0};
    float gyroBias[3] = {0, 0, 0};
    const int samples = 100;
    
    for (int i = 0; i < samples; i++) {
        sensors_event_t a, g, temp;
        mpu.getEvent(&a, &g, &temp);
        
        accelBias[0] += a.acceleration.x;
        accelBias[1] += a.acceleration.y;
        accelBias[2] += a.acceleration.z - 9.81; // Remove gravity
        
        gyroBias[0] += g.gyro.x;
        gyroBias[1] += g.gyro.y;
        gyroBias[2] += g.gyro.z;
        
        delay(10);
    }
    
    // Calculate average biases
    for (int i = 0; i < 3; i++) {
        accelBias[i] /= samples;
        gyroBias[i] /= samples;
    }
    
    Serial.println("Calibration complete");
}

void QuadrupedRobot::setPIDGains(float kp, float ki, float kd) {
    pidGains[0] = kp;
    pidGains[1] = ki;
    pidGains[2] = kd;
}

void QuadrupedRobot::setMaxSpeed(float speed) {
    maxSpeed = constrain(speed, 0.0f, 1.0f);
}

void QuadrupedRobot::setStepHeight(float height) {
    gaitParams.stepHeight = height;
}

void QuadrupedRobot::setBodyHeight(float height) {
    bodyHeight = height;
    stand(); // Update standing position with new height
}

void QuadrupedRobot::setCycleTime(float time) {
    gaitParams.cycleTime = time;
}

void QuadrupedRobot::setLegPhaseOffset(int legIndex, float offset) {
    if (legIndex >= 0 && legIndex < NUM_LEGS) {
        gaitParams.phaseOffsets[legIndex] = offset;
    }
}

void QuadrupedRobot::walk(int cycles, float velocity, Direction direction) {
    if (velocity > maxSpeed) velocity = maxSpeed;
    moving = true;
    
    unsigned long startTime = millis();
    int currentCycle = 0;
    
    while (moving && currentCycle < cycles) {
        unsigned long currentTime = millis();
        float phase = ((currentTime - startTime) % (unsigned long)gaitParams.cycleTime) / gaitParams.cycleTime;
        
        generateWalkingGait(phase, velocity, direction);
        
        float pitchCorrection, rollCorrection;
        calculateBalanceCorrection(pitchCorrection, rollCorrection);
        applyBalanceCorrection(pitchCorrection, rollCorrection);
        
        if (!isBalanced()) {
            stabilize();
            break;
        }
        
        if (currentTime - startTime >= (currentCycle + 1) * gaitParams.cycleTime) {
            currentCycle++;
        }
        
        delay(10);
    }
    
    moving = false;
}

void QuadrupedRobot::stand() {
    moving = false;
    
    for (int leg = 0; leg < NUM_LEGS; leg++) {
        float hipAngle, kneeAngle;
        calculateInverseKinematics(
            legConfigs[leg].x,
            legConfigs[leg].y,
            -bodyHeight,
            hipAngle,
            kneeAngle,
            legConfigs[leg].isMirrored
        );
        
        setServoAngle(leg * 2, hipAngle);
        setServoAngle(leg * 2 + 1, kneeAngle);
        isGrounded[leg] = true;
    }
}

void QuadrupedRobot::balance() {
    float pitchCorrection, rollCorrection;
    calculateBalanceCorrection(pitchCorrection, rollCorrection);
    applyBalanceCorrection(pitchCorrection, rollCorrection);
}

void QuadrupedRobot::stabilize() {
    updateMpuData();
    
    if (abs(mpuData.pitch) > 45.0f || abs(mpuData.roll) > 45.0f) {
        emergencyStop();
        return;
    }
    
    balance();
    
    if (isBalanced()) {
        stand();
    }
}

void QuadrupedRobot::setGait(GaitType gait) {
    currentGait = gait;
    
    switch (gait) {
        case TROT:
            gaitParams.phaseOffsets[0] = 0.0f;
            gaitParams.phaseOffsets[1] = 0.5f;
            gaitParams.phaseOffsets[2] = 0.5f;
            gaitParams.phaseOffsets[3] = 0.0f;
            break;
        case WALK:
            gaitParams.phaseOffsets[0] = 0.0f;
            gaitParams.phaseOffsets[1] = 0.25f;
            gaitParams.phaseOffsets[2] = 0.5f;
            gaitParams.phaseOffsets[3] = 0.75f;
            break;
        case GALLOP:
            gaitParams.phaseOffsets[0] = 0.0f;
            gaitParams.phaseOffsets[1] = 0.1f;
            gaitParams.phaseOffsets[2] = 0.5f;
            gaitParams.phaseOffsets[3] = 0.6f;
            break;
        case BOUND:
            gaitParams.phaseOffsets[0] = 0.0f;
            gaitParams.phaseOffsets[1] = 0.0f;
            gaitParams.phaseOffsets[2] = 0.5f;
            gaitParams.phaseOffsets[3] = 0.5f;
            break;
    }
}

void QuadrupedRobot::jump(float height, float duration) {
    // Prepare for jump
    crouch(bodyHeight * 0.7f);
    delay(200);
    
    // Jump motion
    for (int leg = 0; leg < NUM_LEGS; leg++) {
        float hipAngle, kneeAngle;
        calculateInverseKinematics(
            legConfigs[leg].x,
            legConfigs[leg].y,
            -bodyHeight + height,
            hipAngle,
            kneeAngle,
            legConfigs[leg].isMirrored
        );
        
        setServoAngle(leg * 2, hipAngle);
        setServoAngle(leg * 2 + 1, kneeAngle);
    }
    
    delay(duration);
    
    // Land
    stand();
}

void QuadrupedRobot::turn(float angle) {
    const float turnRadius = sqrt(
        legConfigs[0].x * legConfigs[0].x + 
        legConfigs[0].y * legConfigs[0].y
    );
    
    for (int leg = 0; leg < NUM_LEGS; leg++) {
        float currentAngle = atan2(legConfigs[leg].y, legConfigs[leg].x);
        float newAngle = currentAngle + angle * M_PI / 180.0f;
        
        float hipAngle, kneeAngle;
        calculateInverseKinematics(
            cos(newAngle) * turnRadius,
            sin(newAngle) * turnRadius,
            -bodyHeight,
            hipAngle,
            kneeAngle,
            legConfigs[leg].isMirrored
        );
        
        setServoAngle(leg * 2, hipAngle);
        setServoAngle(leg * 2 + 1, kneeAngle);
    }
}

void QuadrupedRobot::crouch(float height) {
    for (int leg = 0; leg < NUM_LEGS; leg++) {
        float hipAngle, kneeAngle;
        calculateInverseKinematics(
            legConfigs[leg].x,
            legConfigs[leg].y,
            -height,
            hipAngle,
            kneeAngle,
            legConfigs[leg].isMirrored
        );
        
        setServoAngle(leg * 2, hipAngle);
        setServoAngle(leg * 2 + 1, kneeAngle);
    }
}

void QuadrupedRobot::stretch() {
    // Stretch sequence
    for (int leg = 0; leg < NUM_LEGS; leg++) {
        // Lift leg
        float hipAngle, kneeAngle;
        calculateInverseKinematics(
            legConfigs[leg].x,
            legConfigs[leg].y,
            -bodyHeight + 50.0f,
            hipAngle,
            kneeAngle,
            legConfigs[leg].isMirrored
        );
        
        setServoAngle(leg * 2, hipAngle);
        setServoAngle(leg * 2 + 1, kneeAngle);
        
        delay(500);
        
        // Return to standing
        calculateInverseKinematics(
            legConfigs[leg].x,
            legConfigs[leg].y,
            -bodyHeight,
            hipAngle,
            kneeAngle,
            legConfigs[leg].isMirrored
        );
        
        setServoAngle(leg * 2, hipAngle);
        setServoAngle(leg * 2 + 1, kneeAngle);
        
        delay(500);
    }
}

bool QuadrupedRobot::isBalanced() const {
    const float threshold = 5.0f; // degrees
    return abs(mpuData.pitch) < threshold && abs(mpuData.roll) < threshold;
}

bool QuadrupedRobot::isMoving() const {
    return moving;
}

float QuadrupedRobot::getBatteryLevel() const {
    // Implement battery level reading
    return 0.0f;  // Placeholder
}

const MpuData& QuadrupedRobot::getSensorData() const {
    return mpuData;
}

void QuadrupedRobot::emergencyStop() {
    moving = false;
    state = EMERGENCY;
    stand();
    Serial.println("Emergency stop triggered!");
}

void QuadrupedRobot::updateMpuData() {
    sensors_event_t a, g, temp;
    mpu.getEvent(&a, &g, &temp);
    
    mpuData.accel_x = a.acceleration.x;
    mpuData.accel_y = a.acceleration.y;
    mpuData.accel_z = a.acceleration.z;
    
    mpuData.gyro_x = g.gyro.x;
    mpuData.gyro_y = g.gyro.y;
    mpuData.gyro_z = g.gyro.z;
    
    mpuData.temp = temp.temperature;
    
    mpuData.calculateAngles();
}

void QuadrupedRobot::calculateBalanceCorrection(float& pitchCorrection, float& rollCorrection) {
    const float dt = 0.01f; // Fixed time step for stability
    
    float pitch = mpuData.pitch;
    float roll = mpuData.roll;
    
    // Simple P controller for stability
    pitchCorrection = pidGains[0] * pitch;
    rollCorrection = pidGains[0] * roll;
}

void QuadrupedRobot::applyBalanceCorrection(float pitchCorrection, float rollCorrection) {
    for (int leg = 0; leg < NUM_LEGS; leg++) {
        float z = -bodyHeight;
        
        // Apply pitch correction (front/back)
        if (leg < 2) z += pitchCorrection * 20.0f;
        else z -= pitchCorrection * 20.0f;
        
        // Apply roll correction (left/right)
        if (leg % 2) z += rollCorrection * 20.0f;
        else z -= rollCorrection * 20.0f;
        
        float hipAngle, kneeAngle;
        calculateInverseKinematics(
            legConfigs[leg].x,
            legConfigs[leg].y,
            z,
            hipAngle,
            kneeAngle,
            legConfigs[leg].isMirrored
        );
        
        setServoAngle(leg * 2, hipAngle);
        setServoAngle(leg * 2 + 1, kneeAngle);
    }
}

void QuadrupedRobot::generateWalkingGait(float phase, float velocity, Direction direction) {
    float stride = velocity * gaitParams.strideLength;
    
    for (int leg = 0; leg < NUM_LEGS; leg++) {
        float legPhase = fmod(phase + gaitParams.phaseOffsets[leg], 1.0f);
        float x = legConfigs[leg].x;
        float y = legConfigs[leg].y;
        float z = -bodyHeight;
        
        if (legPhase < 0.5f) {
            // Stance phase
            float stanceX = cos(legPhase * M_PI * 2) * stride;
            float stanceY = 0;
            
            switch (direction) {
                case BACKWARD:
                    stanceX = -stanceX;
                    break;
                case LEFT:
                    stanceY = stanceX;
                    stanceX = 0;
                    break;
                case RIGHT:
                    stanceY = -stanceX;
                    stanceX = 0;
                    break;
                default: // FORWARD
                    break;
            }
            
            x += stanceX;
            y += stanceY;
            isGrounded[leg] = true;
            lastContactTime[leg] = millis();
        } else {
            // Swing phase
            z = -bodyHeight + sin((legPhase - 0.5f) * M_PI * 2) * gaitParams.stepHeight;
            isGrounded[leg] = false;
        }
        
        float hipAngle, kneeAngle;
        calculateInverseKinematics(x, y, z, hipAngle, kneeAngle, legConfigs[leg].isMirrored);
        
        setServoAngle(leg * 2, hipAngle);
        setServoAngle(leg * 2 + 1, kneeAngle);
    }
}

void QuadrupedRobot::calculateLegForces() {
    // Implement force calculation for each leg
    // This is a placeholder for actual force sensor implementation
}

void QuadrupedRobot::calculateInverseKinematics(float x, float y, float z, 
                                               float& hipAngle, float& kneeAngle,
                                               bool isMirrored) {
    const float L1 = 50.0f; // Upper leg length
    const float L2 = 50.0f; // Lower leg length
    
    // Mirror the Y coordinate for left legs
    if (isMirrored) {
        y = -y;
    }
    
    // Calculate leg length and angle in the x-y plane
    float L = sqrt(x*x + y*y);
    float gamma = atan2(y, x);
    
    // Calculate leg length in the x-z plane
    float D = sqrt(L*L + z*z);
    
    // Check if position is reachable
    if (D > L1 + L2) {
        float scale = (L1 + L2) / D;
        x *= scale;
        y *= scale;
        z *= scale;
        D = L1 + L2;
    }
    
    // Calculate joint angles using cosine law
    float alpha = acos((L1*L1 + D*D - L2*L2)/(2*L1*D)) + atan2(z, L);
    float beta = acos((L1*L1 + L2*L2 - D*D)/(2*L1*L2));
    
    // Convert to degrees and apply mirroring if needed
    hipAngle = clampAngle(gamma * 180.0f/M_PI, -90.0f, 90.0f);
    kneeAngle = clampAngle((M_PI - beta) * 180.0f/M_PI, 0.0f, 150.0f);
    
    if (isMirrored) {
        hipAngle = mirrorAngle(hipAngle, true);
    }
}

void QuadrupedRobot::setServoAngle(int servoIndex, float angle) {
    if (servoIndex >= TOTAL_SERVOS) return;
    
    const ServoConfig& config = servoConfigs[servoIndex];
    float adjustedAngle = angle + config.offset;
    
    if (config.reversed) {
        adjustedAngle = -adjustedAngle;
    }
    
    if (config.isMirrored) {
        adjustedAngle = mirrorAngle(adjustedAngle, true);
    }
    
    adjustedAngle = clampAngle(adjustedAngle, config.minAngle, config.maxAngle);
    float pulse = map(adjustedAngle, -90, 90, config.minPulse, config.maxPulse);
    
    pwm.setPWM(config.pin, 0, (int)pulse);
    currentLegAngles[servoIndex] = angle;
}

float QuadrupedRobot::clampAngle(float angle, float min, float max) {
    return fmax(min, fmin(max, angle));
}

float QuadrupedRobot::mirrorAngle(float angle, bool isMirrored) {
    return isMirrored ? -angle : angle;
}

void QuadrupedRobot::validateMovement() {
    for (int i = 0; i < NUM_LEGS; i++) {
        // Check leg extension limits
        float x = legConfigs[i].x;
        float y = legConfigs[i].y;
        float z = legConfigs[i].z;
        
        float extension = sqrt(x*x + y*y + z*z);
        if (extension > 100.0f) { // Maximum leg extension
            emergencyStop();
            Serial.println("Leg extension limit exceeded!");
            return;
        }
        
        // Check servo angle limits
        if (abs(currentLegAngles[i*2]) > 90.0f || 
            currentLegAngles[i*2+1] < 0.0f || 
            currentLegAngles[i*2+1] > 150.0f) {
            emergencyStop();
            Serial.println("Servo angle limit exceeded!");
            return;
        }
    }
}

void QuadrupedRobot::optimizeEnergyUsage() {
    // Implement power optimization strategies
    // This is a placeholder for actual power management
}

void QuadrupedRobot::monitorHealth() {
    // Check orientation limits
    if (abs(mpuData.pitch) > 45.0f || abs(mpuData.roll) > 45.0f) {
        emergencyStop();
        return;
    }
    
    // Validate movement and leg positions
    validateMovement();
}

File: src/robot/QuadrupedRobot.js
export class QuadrupedRobot {
  constructor() {
    this.NUM_LEGS = 4;
    this.SERVOS_PER_LEG = 2;
    this.TOTAL_SERVOS = this.NUM_LEGS * this.SERVOS_PER_LEG;
    
    this.currentLegAngles = new Array(this.TOTAL_SERVOS).fill(0);
    this.targetAngles = new Array(this.TOTAL_SERVOS).fill(0);
    this.moving = false;
    this.bodyHeight = 100;
    
    this.sensorData = {
      accel: { x: 0, y: 0, z: 0 },
      gyro: { x: 0, y: 0, z: 0 },
      orientation: { pitch: 0, roll: 0, yaw: 0 }
    };
    
    this.pidGains = { kp: 0.8, ki: 0.2, kd: 0.1 };
    this.pidState = {
      lastPitchError: 0,
      lastRollError: 0,
      pitchErrorSum: 0,
      rollErrorSum: 0,
      lastTime: Date.now()
    };
    
    this.gaitParams = {
      stepHeight: 30,
      strideLength: 50,
      cycleTime: 1000, // ms per cycle
      phaseOffsets: [0, 0.25, 0.5, 0.75]
    };
    
    this.initializeLegs();
  }

  initializeLegs() {
    const legPositions = [
      { x: 100, y: 100 },   // Front Right
      { x: 100, y: -100 },  // Front Left
      { x: -100, y: 100 },  // Back Right
      { x: -100, y: -100 }  // Back Left
    ];

    this.legs = Array(this.NUM_LEGS).fill().map((_, i) => ({
      id: i,
      position: { ...legPositions[i], z: 0 },
      angles: { hip: 0, knee: 0 },
      isGrounded: true,
      lastContact: Date.now(),
      forceVector: { x: 0, y: 0, z: 0 }
    }));
  }

  walk(cycles, velocity, direction) {
    if (!this.validateParams(velocity)) return false;
    
    this.moving = true;
    const startTime = Date.now();
    let currentCycle = 0;
    
    const walkCycle = () => {
      if (!this.moving || currentCycle >= cycles) {
        this.moving = false;
        return;
      }
      
      const elapsed = Date.now() - startTime;
      const phase = (elapsed % this.gaitParams.cycleTime) / this.gaitParams.cycleTime;
      
      this.generateWalkingGait(phase, velocity, direction);
      this.applyBalanceCorrection();
      
      if (!this.isBalanced()) {
        this.stabilize();
        this.moving = false;
        return;
      }
      
      if (elapsed >= (currentCycle + 1) * this.gaitParams.cycleTime) {
        currentCycle++;
      }
      
      requestAnimationFrame(walkCycle);
    };
    
    walkCycle();
    return true;
  }

  validateParams(velocity) {
    if (velocity < 0 || velocity > 1) {
      console.error('Velocity must be between 0 and 1');
      return false;
    }
    return true;
  }

  generateWalkingGait(phase, velocity, direction) {
    const stride = velocity * this.gaitParams.strideLength;
    
    this.legs.forEach((leg, i) => {
      let legPhase = (phase + this.gaitParams.phaseOffsets[i]) % 1;
      let x = 0, y = 0, z = -this.bodyHeight;
      
      if (legPhase < 0.5) {
        // Stance phase - foot in contact with ground
        x = Math.cos(legPhase * Math.PI * 2) * stride;
        leg.isGrounded = true;
        leg.lastContact = Date.now();
      } else {
        // Swing phase - foot in air
        x = Math.cos(legPhase * Math.PI * 2) * stride;
        z = -this.bodyHeight + Math.sin((legPhase - 0.5) * Math.PI * 2) * this.gaitParams.stepHeight;
        leg.isGrounded = false;
      }
      
      // Apply direction modifications
      const directionModifiers = {
        'FORWARD': [x, y],
        'BACKWARD': [-x, y],
        'LEFT': [-y, x],
        'RIGHT': [y, -x]
      };
      
      [x, y] = directionModifiers[direction] || directionModifiers['FORWARD'];
      
      // Add leg's base position
      x += leg.position.x;
      y += leg.position.y;
      
      const angles = this.calculateInverseKinematics(x, y, z);
      this.setLegAngles(i, angles);
      
      // Update leg state
      leg.position = { x, y, z };
      leg.forceVector = this.calculateLegForce(leg, legPhase);
    });
  }

  calculateLegForce(leg, phase) {
    const force = { x: 0, y: 0, z: 0 };
    
    if (leg.isGrounded) {
      const contactTime = Date.now() - leg.lastContact;
      const contactForce = Math.min(1, contactTime / 100); // Ramp up force over 100ms
      
      force.z = -9.81 * this.bodyHeight * 0.25 * contactForce; // Distribute body weight
      
      // Add horizontal forces based on movement
      force.x = (leg.position.x - leg.position.x) * 0.1;
      force.y = (leg.position.y - leg.position.y) * 0.1;
    }
    
    return force;
  }

  calculateInverseKinematics(x, y, z) {
    const L1 = 50; // Upper leg length
    const L2 = 50; // Lower leg length
    
    // Calculate leg length and angle in the x-y plane
    const L = Math.sqrt(x*x + y*y);
    const gamma = Math.atan2(y, x);
    
    // Calculate leg length in the x-z plane
    const D = Math.sqrt(L*L + z*z);
    
    // Check if position is reachable
    if (D > L1 + L2) {
      console.warn('Position out of reach, adjusting...');
      const scale = (L1 + L2) / D;
      x *= scale;
      y *= scale;
      z *= scale;
    }
    
    // Calculate joint angles using cosine law
    const alpha = Math.acos((L1*L1 + D*D - L2*L2)/(2*L1*D)) + Math.atan2(z, L);
    const beta = Math.acos((L1*L1 + L2*L2 - D*D)/(2*L1*L2));
    
    // Convert to degrees and ensure angles are within servo limits
    return {
      hip: this.clampAngle(gamma * 180/Math.PI, -90, 90),
      knee: this.clampAngle((Math.PI - beta) * 180/Math.PI, 0, 150)
    };
  }

  clampAngle(angle, min, max) {
    return Math.max(min, Math.min(max, angle));
  }

  setLegAngles(legIndex, angles) {
    const servoIndex = legIndex * this.SERVOS_PER_LEG;
    this.currentLegAngles[servoIndex] = angles.hip;
    this.currentLegAngles[servoIndex + 1] = angles.knee;
    
    // Emit angle updates if needed
    this.onAngleUpdate?.(legIndex, angles);
  }

  balance() {
    this.updateSensorData();
    const correction = this.calculateBalanceCorrection();
    this.applyBalanceCorrection(correction);
  }

  updateSensorData() {
    // In a real implementation, this would read from actual sensors
    // Here we simulate sensor noise and drift
    const now = Date.now();
    const dt = (now - this.pidState.lastTime) / 1000;
    
    this.sensorData.accel = {
      x: Math.sin(now * 0.001) * 0.1,
      y: Math.cos(now * 0.001) * 0.1,
      z: -9.81 + Math.sin(now * 0.002) * 0.05
    };
    
    this.sensorData.gyro = {
      x: (Math.random() - 0.5) * 0.1,
      y: (Math.random() - 0.5) * 0.1,
      z: (Math.random() - 0.5) * 0.1
    };
    
    this.calculateOrientation(dt);
    this.pidState.lastTime = now;
  }

  calculateOrientation(dt) {
    const alpha = 0.96; // Complementary filter coefficient
    
    // Calculate angles from accelerometer
    const accelPitch = Math.atan2(this.sensorData.accel.x, 
      Math.sqrt(this.sensorData.accel.y * this.sensorData.accel.y + 
                this.sensorData.accel.z * this.sensorData.accel.z));
                
    const accelRoll = Math.atan2(this.sensorData.accel.y, 
      Math.sqrt(this.sensorData.accel.x * this.sensorData.accel.x + 
                this.sensorData.accel.z * this.sensorData.accel.z));
    
    // Complementary filter
    this.sensorData.orientation.pitch = alpha * 
      (this.sensorData.orientation.pitch + this.sensorData.gyro.x * dt) + 
      (1 - alpha) * accelPitch;
      
    this.sensorData.orientation.roll = alpha * 
      (this.sensorData.orientation.roll + this.sensorData.gyro.y * dt) + 
      (1 - alpha) * accelRoll;
      
    this.sensorData.orientation.yaw += this.sensorData.gyro.z * dt;
  }

  calculateBalanceCorrection() {
    const dt = (Date.now() - this.pidState.lastTime) / 1000;
    
    const pitch = this.sensorData.orientation.pitch;
    const roll = this.sensorData.orientation.roll;
    
    // Update error sums for integral term
    this.pidState.pitchErrorSum += pitch * dt;
    this.pidState.rollErrorSum += roll * dt;
    
    // Calculate PID terms
    const pitchCorrection = 
      this.pidGains.kp * pitch +
      this.pidGains.ki * this.pidState.pitchErrorSum +
      this.pidGains.kd * (pitch - this.pidState.lastPitchError) / dt;
      
    const rollCorrection = 
      this.pidGains.kp * roll +
      this.pidGains.ki * this.pidState.rollErrorSum +
      this.pidGains.kd * (roll - this.pidState.lastRollError) / dt;
    
    // Update last error values
    this.pidState.lastPitchError = pitch;
    this.pidState.lastRollError = roll;
    
    return { pitch: pitchCorrection, roll: rollCorrection };
  }

  applyBalanceCorrection(correction = { pitch: 0, roll: 0 }) {
    this.legs.forEach((leg, i) => {
      let z = -this.bodyHeight;
      
      // Apply pitch correction (front/back)
      if (i < 2) z += correction.pitch * 20;
      else z -= correction.pitch * 20;
      
      // Apply roll correction (left/right)
      if (i % 2) z += correction.roll * 20;
      else z -= correction.roll * 20;
      
      const angles = this.calculateInverseKinematics(
        leg.position.x,
        leg.position.y,
        z
      );
      
      this.setLegAngles(i, angles);
    });
  }

  stand() {
    this.moving = false;
    
    // Move all legs to default standing position
    this.legs.forEach((leg, i) => {
      const angles = this.calculateInverseKinematics(
        leg.position.x,
        leg.position.y,
        -this.bodyHeight
      );
      
      this.setLegAngles(i, angles);
      leg.isGrounded = true;
    });
  }

  stabilize() {
    this.updateSensorData();
    
    // If severely unbalanced, emergency stop
    if (Math.abs(this.sensorData.orientation.pitch) > 45 ||
        Math.abs(this.sensorData.orientation.roll) > 45) {
      this.emergencyStop();
      return;
    }
    
    this.balance();
    
    // Check if stabilization is complete
    if (this.isBalanced()) {
      this.stand();
    }
  }

  emergencyStop() {
    this.moving = false;
    this.stand();
    console.warn('Emergency stop triggered');
  }

  isBalanced() {
    const pitchThreshold = 5; // degrees
    const rollThreshold = 5;  // degrees
    
    return Math.abs(this.sensorData.orientation.pitch) < pitchThreshold &&
           Math.abs(this.sensorData.orientation.roll) < rollThreshold;
  }

  getStatus() {
    return {
      moving: this.moving,
      legPositions: this.legs.map(leg => ({
        position: leg.position,
        isGrounded: leg.isGrounded,
        angles: leg.angles,
        force: leg.forceVector
      })),
      orientation: this.sensorData.orientation,
      isBalanced: this.isBalanced(),
      bodyHeight: this.bodyHeight
    };
  }
}